packages/meditrak-server/src/dataAccessors/updateOrCreateSurveyResponse.js

```diff
@@ -45,49 +45,59 @@ async function saveAnswer(models, answer, surveyResponseId) {
  * Creates or updates survey responses from passed changes
  */
 export async function updateOrCreateSurveyResponse(models, surveyResponseObject) {
+  const {
+    id: surveyResponseId,
+    answers,
+    clinic_id: clinicId,
+    ...surveyResponseProperties
+  } = surveyResponseObject;
+  const entitiesCreated = surveyResponseObject.entities_created || [];
+  const optionsCreated = surveyResponseObject.options_created || [];
+  let surveyResponse;
   try {
-    const {
-      id: surveyResponseId,
-      answers,
-      clinic_id: clinicId,
-      ...surveyResponseProperties
-    } = surveyResponseObject;
-    const entitiesCreated = surveyResponseObject.entities_created || [];
-
-    let surveyResponse;
-    try {
-      await createEntities(models, entitiesCreated, surveyResponseObject.survey_id);
-
-      if (clinicId) {
-        const entityId = await getEntityIdFromClinicId(models, clinicId);
-        surveyResponseProperties.entity_id = entityId;
-      }
-      surveyResponse = await models.surveyResponse.updateOrCreate(
-        {
-          id: surveyResponseId,
-        },
-        {
-          id: surveyResponseId,
-          ...surveyResponseProperties,
-        },
-      );
-      if (!surveyResponse.submission_time) {
-        // Survey responses from older versions of Tupaia MediTrak may not have a submission time
-        surveyResponse.submission_time = surveyResponse.end_time;
-        await surveyResponse.save();
-      }
-    } catch (error) {
-      throw new DatabaseError(
-        `creating/updating survey response with id ${surveyResponseId}`,
-        error,
-      );
+    await createEntities(models, entitiesCreated, surveyResponseObject.survey_id);
+    await createOptions(models, optionsCreated);
+    if (clinicId) {
+      const entityId = await getEntityIdFromClinicId(models, clinicId);
+      surveyResponseProperties.entity_id = entityId;
+    }
+    surveyResponse = await models.surveyResponse.updateOrCreate(
+      {
+        id: surveyResponseId,
+      },
+      {
+        id: surveyResponseId,
+        ...surveyResponseProperties,
+      },
+    );
+    if (!surveyResponse.submission_time) {
+      // Survey responses from older versions of Tupaia MediTrak may not have a submission time
+      surveyResponse.submission_time = surveyResponse.end_time;
+      await surveyResponse.save();
     }
-    await Promise.all(answers.map(answer => saveAnswer(models, answer, surveyResponse.id)));
   } catch (error) {
-    throw error; // Throw error up for caller to handle
+    throw new DatabaseError(`creating/updating survey response with id ${surveyResponseId}`, error);
   }
+  await Promise.all(answers.map(answer => saveAnswer(models, answer, surveyResponse.id)));
 }
 
+const createOptions = async (models, optionsCreated) => {
+  return Promise.all(
+    optionsCreated.map(async option => {
```

Comment:
It might be good to process these sequentially. If the app has created two in the same option set, we may end up putting them as the same sort order if this happens in parallel: option a might be being added when option b looks for the largest sort order, and the db doesn't yet know about option a being added.